packages/next-swc/crates/next-core/js/src/entry/router.ts

```diff
@@ -110,71 +107,61 @@ export default async function route(
     // the serverRequest to Next.js to handle.
     clientRequest.end(body)
 
-    // The route promise must not block us from starting the client response
-    // handling, so we cannot await it yet. By making the call, we allow
-    // Next.js to start writing to the response whenever it's ready.
+    // The route promise must not block us from starting the middleware
+    // response handling, so we cannot await it yet. By making the call, we
+    // allow Next.js to start writing to the response whenever it's ready.
     const routePromise = resolveRoute(serverRequest, serverResponse)
 
-    // Now that the Next.js has started processing the route, the
-    // clientResponsePromise will resolve once they write data and then we can
-    // begin streaming.
-    // We again cannot block on the clientResponsePromise, because an error may
+    // Now that the Next.js has started processing the route, the middleware
+    // response promise will resolve once they write data and then we can begin
+    // streaming.
+    // We again cannot await directly on the promise, because an error may
     // occur in the routePromise while we're waiting.
-    const responsePromise = clientResponsePromise.then((c) =>
-      handleClientResponse(ipc, c)
+    const middlewarePromise = clientResponsePromise.then((c) =>
+      handleMiddlewareResponse(ipc, c)
     )
 
     // Now that both promises are in progress, we await both so that a
     // rejection in either will end the routing.
-    const [response] = await Promise.all([responsePromise, routePromise])
+    const [routeResult] = await Promise.all([routePromise, middlewarePromise])
 
     server.close()
-    return response
+
+    if (routeResult) {
+      switch (routeResult.type) {
+        case 'none':
+        case 'error':
+          return routeResult
+        case 'rewrite':
+          return {
+            type: 'rewrite',
+            data: {
+              url: routeResult.url,
+              headers: Object.entries(routeResult.headers)
+                .filter(([, val]) => val != null)
+                .map(([name, value]) => [name, value!.toString()]),
+            },
+          }
+        default:
+          // @ts-expect-error data.type is never
+          throw new Error(`unknown route result type: ${data.type}`)
+      }
+    }
   } catch (e) {
     // Server doesn't need to be closed, because the sendError will terminate
     // the process.
     ipc.sendError(e as Error)
   }
 }
 
-async function handleClientResponse(
+async function handleMiddlewareResponse(
   ipc: Ipc<RouterRequest, IpcOutgoingMessage>,
   clientResponse: IncomingMessage
-): Promise<MessageData | void> {
-  if (clientResponse.headers['x-nextjs-route-result'] === '1') {
-    clientResponse.setEncoding('utf8')
-    // We're either a redirect or a rewrite
-    let buffer = ''
-    for await (const chunk of clientResponse) {
-      buffer += chunk
-    }
-
-    const data = JSON.parse(buffer) as RouteResult
```

Comment:
This was erroring out for HEAD requests, because the `clientResponse` always has an empty body.